### ISSUE TITLE
[1.*] Webpack code splitting - wrong directory

### ISSUE THREAD

**Issue by KKSzymanowski** [JeffreyWay/laravel-mix#936]
- Laravel Mix Version: 1.0.6
- Node Version: 8.1.0 
- NPM Version: 5.0.3
- Yarn Version: 0.21.3
- OS: Windows 10

### Description:
With Laravel Mix 0.* modules created with webpack code splitting were automatically put in the `public/js` directory. Now, with Mix 1.* it puts all the chunks in the `public` directory(unless for each chunk I specify its name with `js/` prefix).

### Steps To Reproduce:
- `$ laravel new foo --dev`
- Update Mix version (`0.*` to `1.*`) in `package.json`
- `$ yarn`
- `$ yarn add babel-preset-es2015 babel-preset-stage-2` (not sure here which one gives the ability to import *dynamically*)
- Create `.babelrc` file:
```js
{
  "presets": ["es2015", "stage-2"]
}
``` 
- In `app.js` split the Example component to a separate chunk
```js
// Replace this
Vue.component('example', require('./components/Example.vue'));
// With this
Vue.component('example', () => import('./components/Example.vue'));
```
- `$ yarn run development`

The `0.js` file should appear in `public` directory.

### Possible fix:
In `webpack.mix.js` add
```js
mix.webpackConfig({
    output: {
        chunkFilename: 'js/[name].[chunkhash].js',
    },
});
```
Is there a better way to fix this?

**Comment by jbbr** [comment]
I can confirm this behavior.
Path behavior seems to be broken (or at least very different) in 1.x. This might be related to #923 and #924

**Comment by KKSzymanowski** [comment]
@jbbr I'm not sure it is related. Previously, in 0.*, when I dumped the webpack config, the `chunkFilename` was set to:
```
'js/[name].[chunkhash].js'
```
It was probably removed by mistake.

**Comment by jbbr** [comment]
okay, not sure. I'll try to fix this and the publicPath issue I experience once I find the time.

If anyone wants to try a fix:
Relevant line in 1.12.x: https://github.com/JeffreyWay/laravel-mix/blob/712d4daf07468177eaef4d7f95672513550e914d/src/Mix.js#L59 
Relevant line in 1.0.7: https://github.com/JeffreyWay/laravel-mix/blob/1da78b6ea0c1b256c582f5dbdea9b14802699e50/src/builder/WebpackConfig.js#L67

**Comment by KKSzymanowski** [comment]
Yeah, the 1.* version is missing the base path for chunk names, which was previously derived from the script output path. If the compilation was defined as follows:
```js
mix.js('resources/assets/js/app.js', 'public/js');
```
`global.entry.base` was `public/js`, `output.chunkFilename` was `public/js/[name].[chunkhash].js` and everything worked correctly.
@JeffreyWay Should I make a PR fixing this?

---

Also:
> Relevant line in 1.12.x:

I think you meant 0.12.x

**Comment by jon-heller** [comment]
I believe I'm also having this problem as well. Using WordPress, which has its assets in `/wp-content/themes/[theme name]`, so I need to set the publicPath. 

Oddly, when I tried to manually update it on https://github.com/JeffreyWay/laravel-mix/blob/1da78b6ea0c1b256c582f5dbdea9b14802699e50/src/builder/WebpackConfig.js#L67 as mentioned above, it broke the compliation. Seems other parts of the script try to reference the publicPath for the actual file instead of just the path?

**Comment by KKSzymanowski** [comment]
@jon-heller I believe you can use `webpackModuleName` magic comment to point to the right directory if you want to do this only a couple of times.

**Comment by jon-heller** [comment]
Well, scratch all that! I think I had some confusion about `path` vs `publicPath`. The following configuration is working for me:

webpack.mix.js
```
.webpackConfig({
        output: {
            publicPath: '/wp-content/themes/theme-name/',
            chunkFilename: 'assets/dist/[name].[chunkhash].js',
        },
    })
```

app.js
```
if ($('body').hasClass('home')) {
        require.ensure(
            ['./homepage'],
            require => {
                const Homepage = require('./homepage').default;
                this.homePage = new Homepage();
            },
            'homepage'
        );
    }
```

**Comment by jon-heller** [comment]
Well, I'm confused. Went back and forth a few times but this does seem to be working, and updating the publicPath correctly where it uses this directory as the base URL when looking for chunks.

```
.webpackConfig({
        output: {
            publicPath: '/wp-content/themes/theme-name/',
            chunkFilename: 'assets/dist/[name].[chunkhash].js',
        },
    })
```

**Comment by mtx-z** [comment]
I'm able to split my JS file using `System.import` instead of `require`. 
This works perfectly when running `dev `or `production `from `npm`. But this is not working on `run watch`. 

On file update, I see `npm `making stuff, and outputing splitted file in` public/js` folder. But I think that because of the changing `chunkhash`, the JS file linked in front is the wrong one: still the old (from first pass of `npm run watch`).

If i remove the `chunkhash `from `chunkFilename `parameter, it works, as the file is overrided without any hash in filename; but i loose the browser caching invalidation as filename/url is the same, and `mix.version()` seems to not apply on those files.

**Comment by silvioiannone** [comment]
This worked for me:

    mix.webpackConfig({
        output: {
            chunkFilename: 'js/[name].js',
        }
    });

**Comment by mtx-z** [comment]
@silvioiannone yes, but dont you loose browser header cache invalidation ? As filename is always the same, and mix.version() don't seems to apply on chunked files ?

I finally made a condition "if production" to use chunkFilenames with chunkHash, and without chunkhash for production build. This way, i have my watch working, and my fillename have hashes on production.

**Comment by silvioiannone** [comment]
@mtx-z You're right, didn't think about that, but at the moment we're still working on our project and hopefully this issue will be solved before going live.

**Comment by mtx-z** [comment]
All right, don't forget it !
In the meantime, this is how i managed it; 

```
const {mix} = require('laravel-mix');
if (mix.inProduction()) {
    mix.webpackConfig(... chunkFilename: 'js/[name].[chunkhash].app.js...);
} else {
    mix.webpackConfig(...chunkFilename: 'js/[name].app.js'...);
}
```

**Comment by KKSzymanowski** [comment]
I personally always use versioning. It doesn't give much overhead and it saves me from forgetting to apply the versioning when releasing the app.

**Comment by mtx-z** [comment]
How do you make versionning work fur chunked files ?

**Comment by silvioiannone** [comment]
@KKSzymanowski I switched to your solution so it includes the chunkhash too now, thanks :)

**Comment by mtx-z** [comment]
@KKSzymanowski I'll test. So your solution also work using "npm watch" ? Good chunk hash filename is loaded on front after watch detected a change and built ?

**Comment by silvioiannone** [comment]
@mtx-z It works but you end up with multiple versions of the same chunk:

[Result screenshot](http://imgur.com/XJSJQx9)

**Comment by mtx-z** [comment]
@silvioiannone I have the same, but on the frontend side, after you refreshed your page, does the frontend have the latest file linked into the DOM ? As for me, only the first one created after i ran the npm watch is used. Other files with new chunk are created, but never linked on frontend.

**Comment by silvioiannone** [comment]
@mtx-z Yes, I just made the test because I was unsure and it appears the the newest chunks are always linked.

**Comment by KKSzymanowski** [comment]
I'm no expert in webpack but from what I understand Laravel only provides the main chunk and all others are linked by webpack so you should always get the correct version.

**Comment by mtx-z** [comment]
@silvioiannone hum ok. Strange as here I always get the first file (so first chunkhash filename) linked, new ones are never used until a new npm watch/dev/prod. I'll do more test ! Thanks a lot for your testings.

**Comment by mtx-z** [comment]
@silvioiannone @KKSzymanowski I just tested again. 

My Webpach have filename: `'[name].js', chunkFilename: 'js/[name].[chunkhash].js'`

- `npm watch` generate this file `37.c18fba44997ac048aa8e.js`
- the file is linked in frontend
- now i edit the related compenent
- npm is detecting watch, building and generate `37.0fb2c17e4dabf1e798ee.js`
- refresh browser (without cache checked)
- first file `37.c18fba44997ac048aa8e.js i`s still linked to DOM.

**Comment by silvioiannone** [comment]
What version of Laravel-Mix do you have exactly? Mine is **1.2.2**.

**Comment by mtx-z** [comment]
`^1.0.6` for me. I try update.

**Comment by mtx-z** [comment]
Now at `1.2.2` (had to followed [this ](https://github.com/nuxt/nuxt.js/issues/1084)), and still the same...

**Comment by santigarcor** [comment]
I can confirm this in 1.4.2

**Comment by sausin** [comment]
it would be awesome if this could work with query strings!

In the meantime, @mtx-z's solution seems to work alright

**Comment by geodeveloper** [comment]
Hi there,

I'm having some troubles with this too, I'm using a different path for my public path when running on production, the chunks are generated correctly on their path but on the browser in the production site is loading from my development path. Here you have an example:

```javascript
let mix = require('laravel-mix');

if (mix.inProduction()) {
    mix.setPublicPath('public/build');
}

mix.webpackConfig({
    output: {
        chunkFilename: mix.inProduction() ? 'chunks/[name].[chunkhash].js' : 'chunks/[name].js'
    }
}
```
Chunks generated for development:
`public/chunks/0.0.js`
`public/chunks/1.1.js`

Chunks generated for production:
`public/build/chunks/0.390c52103f7e07f649e0.js`
`public/build/chunks/1.72a4a93750363d33617e.js`

So far so good. Now in production website when the chunks are loaded they load from:
`/chunks/0.390c52103f7e07f649e0.js`
`/chunks/1.72a4a93750363d33617e`

... instead from:
`/build/chunks/0.390c52103f7e07f649e0.js`
`/build/chunks/1.72a4a93750363d33617e.js`

Am I missing something?

**Comment by KKSzymanowski** [comment]
@geodeveloper How about
```diff
let mix = require('laravel-mix');

if (mix.inProduction()) {
    mix.setPublicPath('public/build');
}

mix.webpackConfig({
    output: {
--      chunkFilename: mix.inProduction() ? 'chunks/[name].[chunkhash].js' : 'chunks/[name].js'
++      chunkFilename: mix.inProduction() ? 'build/chunks/[name].[chunkhash].js' : 'chunks/[name].js'
    }
}
```

**Comment by geodeveloper** [comment]
@KKSzymanowski thanks for the reply. I tried your solution but still now working properly, it generates a new nested `build` folder for the chunks, like this:
```
-- public
  -- build
    -- build
       -- chunks
         -- 0.390c52103f7e07f649e0.js
         -- 1.72a4a93750363d33617e.js
```
Tha's because my build folder is set with `mix.setPublicPath('public/build');` so everything is generated within that folder.

**Comment by KKSzymanowski** [comment]
I'm confused. 
In development you want following structure:
```
└─ public
   ├─ chunks
   │  ├─ 0.js
   │  └─ 1.js
   ├─ js
   │  └─ app.js
   └─ mix-manifest.json
```
And in production:
```
└─ public
   ├─ build
   │  └─ chunks
   │     ├─ 0.390c52103f7e07f649e0.js
   │     └─ 1.72a4a93750363d33617e.js
   ├─ js
   │  └─ app.js
   └─ mix-manifest.json
```

Am I correct?

**Comment by geodeveloper** [comment]
Ok, I got it working. I needed to specify where I have the public files in webpack configuration, by adding `publicPath`. Now it seems it loads the chunks correctly from `/build/chunks/` on production site:
```javascript
let mix = require('laravel-mix');

if (mix.inProduction()) {
    mix.setPublicPath('public/build');
}

mix.webpackConfig({
    output: {
         chunkFilename: mix.inProduction() ? 'chunks/[name].[chunkhash].js' : 'chunks/[name].js',
         publicPath: mix.inProduction() ? '/build/' : '/'
    }
}
```

**Comment by Tarasovych** [comment]
Same problem here, laravel-mix@1.7.2

`webpack.mix.js`:
```
mix.webpackConfig({
    output: {
        chunkFilename: 'js/[name].[chunkhash].js',
    },
});
```
results id instead name anyway.

**Comment by zackkrida** [comment]
For anyone still dealing with this, the easiest thing I've found is to _simply let_ the js chunk files float around in the root `/public` directory, you can ignore them from git via the following lines added to your `.gitignore`:

```gitignore
/public/*.js
/public/*.js.map
```

**Comment by TemaSM** [comment]
Related to https://github.com/JeffreyWay/laravel-mix/issues/1889
In the meantime, commenting out `mix.extract()` helped me (v4.0.15)

**Comment by JSn1nj4** [comment]
So, I've tried some suggestions above, namely not having the chunkFilename say `js/modules/[name].js` since that was causing me to end up with this structure:
```
public/
  js/
    modules/
      js/
        named-js-file-1.js
        named-js-file-2.js
        named-js-file-3.js
        ..
    0.js
    1.js
    2.js
    ...
```
By removing that, the structure is now like this:
```
public/
  js/
    named-js-file-1.js
    named-js-file-2.js
    named-js-file-3.js
    ...
  0.js
  1.js
  2.js
  ..
```
However, my app.css file in `public/css/app.css` is full of JS locale references for Webpack instead of CSS (as it was before I removed the `chunkFilename` option). This has forced me to revert back to laravel-mix 3. Has anyone else noticed this particular issue, and are there any recommendations to fix it?
